Add e2e test for mount propagation

jsafrane opened the pull request:
**What this PR does / why we need it**:
This adds e2e test for mount propagation introduced by #46444.

@kubernetes/sig-node-pr-reviews 
/sig node

**Release note**:
```release-note
None
```

yujuhong commented:
> I cannot use tempSetCurrentKubeletConfig as the other tests because it enables features only in kubelet and not in API server.

Not sure if node e2e supports this right now. Most likely you'd need to plumb it through yourself, but others may know more.

On the other hand, there are cluster e2e suites testing all alpha features: 
https://k8s-testgrid.appspot.com/google-gce#gce-alpha-features
https://k8s-testgrid.appspot.com/google-gce#gci-gce-alpha-features

You can start by adding the test there. BTW, One of them seems to be failing because of this feature (#51961).

yujuhong commented:
> Not sure if node e2e supports this right now. Most likely you'd need to plumb it through yourself, but others may know more.

Minor correction: it's possible to pass TEST_ARGS in a new suite ([example](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties#L7)), but I'm not sure about changing it dynamically.

dashpole commented:
Is there a reason why we want to enable it dynamically rather than just add it to whatever test suite?  Will having this enabled cause other tests to fail?

jsafrane commented:
Moved from node_e2e to e2e. What's the procedure for enabling `[Feature:MountPropagation]` in gci-gce-alpha-features? Who gets notifications about broken test runs there?

Note that I left the previous commit there just in case I want to move back to node_e2e. I will squash it before merge.

jsafrane commented:
Squashed, rebased and removed WIP. Ready for review.

It's e2e test now and will be enabled in ci-kubernetes-e2e-gci-gce-alpha-features when this is merged: https://github.com/kubernetes/test-infra/pull/5335

jsafrane commented:
/retest

jsafrane commented:
Rebased

tallclair commented on test/e2e/node/mount_propagation.go:
```diff
@@ -0,0 +1,178 @@
+		for _, podName := range podNames {
+			for _, dirName := range podNames {
+				cmd := fmt.Sprintf("test -d /mnt/test/%s", dirName)
+				_ = f.ExecShellInPod(podName, cmd)
```
Are you sure this actually captures a non-zero exit code? (try flipping the test and verify that it fails)

jsafrane replied:
Yes, it leads to [ExecCommandInContainer](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/exec_util.go#L104) that prints error:

```
• Failure [36.984 seconds]
[k8s.io] [sig-node] Mount propagation [Feature:MountPropagation]
/home/jsafrane/project/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:642
  should propagate mounts to the host [It]
  /home/jsafrane/project/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/node/mount_propagation.go:80

  failed to execute command in pod master, container cntr: command terminated with exit code 1
  Expected error:
      <exec.CodeExitError>: {
          Err: {
              s: "command terminated with exit code 1",
          },
          Code: 1,
      }
      command terminated with exit code 1
  not to have occurred
```

tallclair commented on test/e2e/node/mount_propagation.go:
```diff
@@ -0,0 +1,178 @@
+		// Each pod mounts one tmpfs to /mnt/test/<podname> and puts a file there.
+		for _, podName := range podNames {
+			cmd := fmt.Sprintf("mount -t tmpfs e2e-mount-propagation-%[1]s /mnt/test/%[1]s; echo %[1]s > /mnt/test/%[1]s/file", podName)
+			_ = f.ExecShellInPod(podName, cmd)
```
nit: don't need `_ = `

jsafrane replied:
removed

jsafrane commented on test/e2e/node/mount_propagation.go:
```diff
@@ -0,0 +1,178 @@
+		// running in parallel.
+		hostDir := "/var/lib/kubelet/" + f.Namespace.Name
+		defer func() {
+			cleanCmd := fmt.Sprintf("sudo rm -rf %q", hostDir)
```
Umm, I could use `/var/lib/kubelet/tmp/<namespace>` and remove whole `/var/lib/kubelet/tmp/` in the end. How likely is that the same test runs twice in parallel in the same cluster?

jsafrane replied:
I added `if len(f.Namespace.Name) == 0 { error & return }` with some comment around for now.

jsafrane commented on test/e2e/node/mount_propagation.go:
```diff
@@ -0,0 +1,178 @@
+		// hostDir is the directory that's shared via HostPath among all pods.
+		// Make sure it's random enough so we don't clash with another test
+		// running in parallel.
+		hostDir := "/var/lib/kubelet/" + f.Namespace.Name
```
I'd rather not.  https://github.com/kubernetes/kubernetes/pull/45724 makes sure that at least /var/lib/kubelet can be mounted as shared into containers. This is useful on Debian Wheezy and other non-systemd distros, where / is not shareable.

By putting it into /var/lib/kubelet/ we test also #45724.

jsafrane commented on test/e2e/node/mount_propagation.go:
```diff
@@ -0,0 +1,178 @@
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
```
removed

jsafrane commented on test/e2e/node/mount_propagation.go:
```diff
@@ -0,0 +1,178 @@
+
+		podClient := f.PodClient()
+		master := podClient.CreateSync(preparePod("master", node, v1.MountPropagationBidirectional, hostDir))
+		defer podClient.Delete(master.Name, nil)
```
removed, still it's feels like a good practice for a test to clean up its objects.

jsafrane committed:
Review fixes

jsafrane committed:
Add e2e test for mount propagation

jsafrane commented on test/e2e/node/mount_propagation.go:
```diff
@@ -0,0 +1,178 @@
+		// can check mount propagation from the host to pods.
+		cmd := fmt.Sprintf("sudo mkdir %[1]q/host; sudo mount -t tmpfs e2e-mount-propagation-host %[1]q/host; echo host > %[1]q/host/file", hostDir)
+		err := framework.IssueSSHCommand(cmd, framework.TestContext.Provider, node)
+		Expect(err).NotTo(HaveOccurred())
```
fixed

jsafrane commented:
@tallclair, feedback adressed, PTAL

tallclair commented:
/lgtm
